File: packages/providers/src/grok.ts
import type { ProviderUsage } from "./types.js";
import { clampPercent, makePercentWindow, safeFetch } from "./utils.js";

interface GrokBillingCredits {
  config?: {
    creditUsagePercent?: number;
    currentPeriod?: { end?: string };
  };
  productUsage?: Array<{
    name?: string;
    usagePercent?: number;
  }>;
}

export async function fetchGrokUsage(
  accessToken?: string
): Promise<ProviderUsage> {
  const base: ProviderUsage = {
    id: "grok",
    name: "Grok",
    status: "disconnected",
    windows: [],
    lastSyncedAt: new Date().toISOString(),
    accent: "#111827",
  };

  if (!accessToken) {
    return { ...base, error: "No Grok auth token found" };
  }

  const headers = {
    Authorization: `Bearer ${accessToken}`,
    "X-XAI-Token-Auth": "xai-grok-cli",
  };

  const billing = await safeFetch<GrokBillingCredits>(
    "https://cli-chat-proxy.grok.com/v1/billing?format=credits",
    { headers }
  );

  if (!billing.ok) {
    return { ...base, status: "error", error: billing.error };
  }

  const windows: ProviderUsage["windows"] = [];
  const config = billing.data.config;

  if (config?.creditUsagePercent != null) {
    windows.push(
      makePercentWindow(
        "weekly",
        "Weekly credits",
        config.creditUsagePercent,
        config.currentPeriod?.end
      )
    );
  }

  for (const product of billing.data.productUsage ?? []) {
    if (product.usagePercent == null || !product.name) continue;
    windows.push({
      kind: "weekly",
      label: product.name,
      used: clampPercent(product.usagePercent),
      limit: 100,
      remaining: Math.max(0, 100 - clampPercent(product.usagePercent)),
      percentUsed: clampPercent(product.usagePercent),
      resetsAt: config?.currentPeriod?.end,
    });
  }

  if (windows.length === 0) {
    return {
      ...base,
      status: "error",
      error: "No billing windows returned",
    };
  }

  return {
    ...base,
    status: "connected",
    plan: "SuperGrok / Build",
    windows,
  };
}

File: apps/desktop/src-tauri/src/providers/grok.rs
use serde_json::Value;

use super::window_percent;
use crate::types::ProviderUsage;

pub async fn fetch(token: Option<String>) -> ProviderUsage {
    let base = ProviderUsage {
        id: "grok".into(),
        name: "Grok".into(),
        plan: None,
        status: "disconnected".into(),
        windows: vec![],
        last_synced_at: chrono::Utc::now().to_rfc3339(),
        error: None,
        accent: "#111827".into(),
    };

    let token = match token {
        Some(t) if !t.is_empty() => t,
        _ => return ProviderUsage { error: Some("Not connected".into()), ..base },
    };

    let client = reqwest::Client::new();
    let resp = client
        .get("https://cli-chat-proxy.grok.com/v1/billing?format=credits")
        .header("Authorization", format!("Bearer {}", token))
        .header("X-XAI-Token-Auth", "xai-grok-cli")
        .send()
        .await;

    let mut windows = vec![];

    if let Ok(resp) = resp {
        if let Ok(json) = resp.json::<Value>().await {
            if let Some(config) = json.get("config") {
                let pct = config
                    .get("creditUsagePercent")
                    .and_then(|v| v.as_f64())
                    .unwrap_or(0.0);
                let reset = config
                    .pointer("/currentPeriod/end")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string());
                windows.push(window_percent("weekly", "Weekly credits", pct, reset));
            }
        }
    }

    if windows.is_empty() {
        return ProviderUsage {
            status: "error".into(),
            error: Some("Could not read Grok usage. Run `grok login` first.".into()),
            ..base
        };
    }

    ProviderUsage {
        status: "connected".into(),
        plan: Some("SuperGrok".into()),
        windows,
        ..base
    }
}
